clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -989,6 +989,18 @@ static mlir::Value maybeBuildArrayDecay(mlir::OpBuilder &builder,
   assert(arrayPtrTy && "expected pointer type");
   auto arrayTy = arrayPtrTy.getPointee().dyn_cast<::mlir::cir::ArrayType>();
 
+  // FIXME(cir): the check below is rather unintuitive and may not work in all
+  // scenarios. The issue is that this method has no way to check if the
+  // subscript operator is being applied to a pointer or an array. In the first
```

Comment:
How so? Why can't it look through the operation until you find the original type?

Reply:
@bcardosolopes I'm not sure if I understand the suggestion.

If we allocate a local array, `cir.alloca` will store it and return a pointer to an array, and when we pass an array as an argument, it is decayed and received by the function as a pointer to an array. While the first should be decayed, the second, should not, but both have the same type.

By _"original type"_ do you mean looking at the input type of the operation that defined the value to be decayed? E.g. instead of looking at the result type of `cir.alloca` we would look at the input type.

Reply:
I'm a bit confused, too many indirections with words hehe. Can you provide an sketch example of the problem in CIR so it's easier to discuss?

Reply:
Sure! Take the following codegen as an example:

```mlir
// void localArray(void) { int arr[4][4]; arr[2][3]; }
cir.func @localArray() {
  %0 = cir.alloca !cir.array<!cir.array<!s32i x 4> x 4>, cir.ptr <!cir.array<!cir.array<!s32i x 4> x 4>>, ["arr"] {alignment = 16 : i64}
  %1 = cir.const(#cir.int<2> : !s32i) : !s32i
  %2 = cir.cast(array_to_ptrdecay, %0 : !cir.ptr<!cir.array<!cir.array<!s32i x 4> x 4>>), !cir.ptr<!cir.array<!s32i x 4>>
  %3 = cir.ptr_stride(%2 : !cir.ptr<!cir.array<!s32i x 4>>, %1 : !s32i), !cir.ptr<!cir.array<!s32i x 4>>
  %4 = cir.const(#cir.int<3> : !s32i) : !s32i
  %5 = cir.cast(array_to_ptrdecay, %3 : !cir.ptr<!cir.array<!s32i x 4>>), !cir.ptr<!s32i>
  %6 = cir.ptr_stride(%5 : !cir.ptr<!s32i>, %4 : !s32i), !cir.ptr<!s32i>
  %7 = cir.load %6 : cir.ptr <!s32i>, !s32i
  cir.return
}
// void decayedArray(int arr[][4]) { arr[2][3]; }
cir.func @decayedArray(%arg0: !cir.ptr<!cir.array<!s32i x 4>>) {
  %8 = cir.alloca !cir.ptr<!cir.array<!s32i x 4>>, cir.ptr <!cir.ptr<!cir.array<!s32i x 4>>>, ["arr", init] {alignment = 8 : i64}
  cir.store %arg0, %8 : !cir.ptr<!cir.array<!s32i x 4>>, cir.ptr <!cir.ptr<!cir.array<!s32i x 4>>>
  %9 = cir.load %8 : cir.ptr <!cir.ptr<!cir.array<!s32i x 4>>>, !cir.ptr<!cir.array<!s32i x 4>>
  %10 = cir.const(#cir.int<2> : !s32i) : !s32i
  %11 = cir.cast(array_to_ptrdecay, %9 : !cir.ptr<!cir.array<!s32i x 4>>), !cir.ptr<!s32i>
  %12 = cir.ptr_stride(%11 : !cir.ptr<!s32i>, %10 : !s32i), !cir.ptr<!s32i>
  [...]
}
```

In `@decayedArray`, the `%11` should not happen before the stride for two reasons. First, the array is already decayed. Second, because the `%11` decay drops the array size information, which is essential for us to know the size of the stride.

So the question is: when executing `maybeBuildArrayDecay`, how do we know if the pointer in `arrayPtr` is an array that has already been decayed?

My answer to this was to check if `arrayPtr` is the result of a `cir.load`, because, due to how codegen currently works, pointers are always loaded after a subscript operator is applied.

We could look at `arrayPtr`'s type, but I'm not sure it would help: both `%3` and `%9` are pointers to an array, but only the latter should decay.

Reply:
> In @decayedArray, the %11 should not happen before the stride for two reasons. First, the array is already decayed. 

My understanding is that only half of it is decayed, just like after the first decay in `@localArray()`.

> Second, because the %11 decay drops the array size information, which is essential for us to know the size of the stride.

Not sure I understand, when I designed single dimension arrays the idea was to operate on top of the underlying types, I didn't account for operating on top of ptrs to cir.array's, like in `%3 = cir.ptr_stride(%2 : !cir.ptr<!cir.array<!s32i x 4>>`. I also didn't implemented multi-dim arrays so take this with a grain of salt: perhaps the multi-dim array decay should be done all at once and the ptr_stride have an additional operand for the second index? It's fine to go thought multiple hops when designing, we often realize the design isn't a good fit when more pieces start showing up. 

> So the question is: when executing `maybeBuildArrayDecay`, how do we know if the pointer in `arrayPtr` is an array that has already been decayed? My answer to this was to check if `arrayPtr` is the result of a `cir.load`, because, due to how codegen currently works, pointers are always loaded after a subscript operator is applied.

Relying on loads sounds a bit brittle to me, looks like we might need either (a) some extra bookkeeping during codegen, (b) breaking into different builders like you suggested earlier or (c) differentiate bound versus unbound mult-dim arrays in CIR and use that info.

Reply:
Looking at https://github.com/llvm/clangir/blob/main/clang/lib/CIR/CodeGen/CIRGenExpr.cpp#L1123 we put the decay independent of the `ptrstride` and the `GEP` just knows how to do it: https://github.com/llvm/clangir/blob/main/clang/lib/CodeGen/CGExpr.cpp#L3677.

That's probably the better approach -- just have `ptrstride` know how to "cast" the array to a pointer for indexing.

Reply:
> My understanding is that only half of it is decayed

In the second example, the array is passed to the function _"half decayed"_ and it is fully decayed before the first stride. Whereas in the first example, the array takes the first stride  _"half decayed"_, then becomes fully decayed for the second stride.

> when I designed single-dimension arrays the idea was to operate on top of the underlying types, I didn't account for operating on top of ptrs to cir.array's

I'm a bit confused here. You didn't account for when the underlying type is a `!cir.array` or was `ptr_stride` never meant to be used with pointers to arrays?

>  also didn't implemented multi-dim arrays so take this with a grain of salt: perhaps the multi-dim array decay should be done all at once and the ptr_stride have an additional operand for the second index?

It could work, though we would have to track the array's dimensions somehow so that we can calculate the length of the stride. Might be a bit cumbersome.

> (b) breaking into different builders like you suggested earlier

I believe this is the easiest and most concise approach for now.

> just have ptrstride know how to "cast" the array to a pointer for indexing.

I would like something like that too, but with a different operation. Something like a `cir.subscript` that would know how to index both arrays and pointers. If it is too high-level for some passes, we can eventually lower it to a combination of `ptrstride`s and `load`s.

Reply:
> You didn't account for when the underlying type is a !cir.array or was ptr_stride never meant to be used with pointers to arrays?

The second, it's more about "nothing thinking ahead into multi-arrays" than "never meant", I'm very open to better designs or better semantics if you have suggestions.

> I would like something like that too, but with a different operation. Something like a cir.subscript that would know how to index both arrays and pointers. If it is too high-level for some passes, we can eventually lower it to a combination of ptrstrides and loads.

I'm curious to see your `cir.subscript` proposal whenever you have time to propose it, I'm fine with you doing (b) for now to get unblocked though. It's probably best if it comes after stuff is working, so tests can "just pass".

Reply:
@bcardosolopes since (b) would duplicate too much code, I did something a bit different: I added a flag to all relevant methods telling if the access should (or not) decay the base pointer.

This flag is set according to the base type: if it is an array, it should decay, if it is a pointer, it shouldn't.

Reply:
Seems like a bit overkill and perhaps a bit brittle? I imagine we could have several function arguments, with different array subscript levels, etc. Isn't there any bookkeeping that can be done during codegen to facilitate such decisions? TBH, seems like this is asking for better IR design sooner than later...

Reply:
> Seems like a bit overkill and perhaps a bit brittle?

@bcardosolopes I don't think so, for the reasons below.

> I imagine we could have several function arguments, with different array subscript levels, etc. Isn't there any bookkeeping that can be done during codegen to facilitate such decisions?

I believe we can, but regardless, my understanding is that every subscript expression calls `buildArraySubscriptPtr` eventually. When we force a non-optional argument here, we ensure this will have to be specified for any subscript access, which does not seem brittle. Maybe overkill, but certainly less overkill than duplicating the functions.

> TBH, seems like this is asking for better IR design sooner than later...

Should we discuss refactoring this to a new `cir.subscript` operation then?

Reply:
> Maybe overkill, but certainly less overkill than duplicating the functions.

Can you paste an example? Perhaps that would help.

> Should we discuss refactoring this to a new cir.subscript operation then?

I'd like that, specially if that would simplify the amount and quality of CIR we emit. I understand if you have some time constraints because GSoC is almost at the end though.